### DISCUSSION TITLE
unzipping to arbitrary location within /tmp

### DISCUSSION THREAD

**Discussion by ws1088** [pex-tool/pex#1374] [Q&A]
Because I need to read a datafile, I assume I have to use `--not-zip-safe` and therefore when I run my pex-ed file it will unzip to `~/.pex`. Is there a way to instruct pex to unzip to a randomly generated directory within /tmp? i.e. I don't need to cache the unzipped wheels.

**Comment by jsirois** [comment]
First to your assumption: it's not correct; you don't actually need to use `--not-zip-safe`.
To access files uniformly between zipped `sys.path` entries and directory `sys.path` entries you only need to use either:
1. In the standard library: https://docs.python.org/3/library/pkgutil.html#pkgutil.get_data
2. In `setuptools` using `pkg_resources` `resource_stream` or `resource_string`: https://setuptools.readthedocs.io/en/latest/pkg_resources.html#basic-resource-access

If - for whatever reason - you do want to use the traditional file system APIs (`open`), you can control the unzip location via the `PEX_ROOT` environment variable, but that directory will not be automatically cleaned up; so you'd need some scripting around the execution of the PEX. For example:
```bash
#!/usr/bin/env bash

set -euo pipefail

export PEX_ROOT=$(mktemp -d)
trap "rm -rf ${PEX_ROOT}" EXIT
/path/to/my/pex "$@"
```

**Reply by jsirois** [reply]
If you need a more pointed example of how to access resources via `pkgutil` or `pkg_resources` after trying and failing to use them, let me know and I can whip that up. When possible, I'd recommend that approach. Only if 3rdparty code needs to access the data file and expects a filesystem path should you need to resort to `--not-zip-safe`.
